File: acmOtherMaterials/Ulm Local Contest/2002/solution/diplomatic.hs
-- Problem   Diplomatic License
-- Algorithm Geometry
-- Runtime   O(n)
-- Author    Walter Guttmann
-- Date      30.05.2002

import Complex

main :: IO ()
main = readFile "diplomatic.in" >>=
       mapM_ (putStrLn . pretty . solve . parse) . cases . map read . words

type Case  = [Point]
type Point = Complex Double

cases :: [Double] -> [[Double]]
cases [] = []
cases (d:xs) = let (ys,zs) = splitAt (2 * round d) xs in ys : cases zs

parse :: [Double] -> Case
parse [] = []
parse (x:y:xs) = x:+y : parse xs

solve :: Case -> Case
solve ps = zipWith (\x y -> (x + y) / 2.0) ps (tail ps ++ [head ps])

pretty :: Case -> String
pretty ps = show (length ps) ++
            concat [ " " ++ show x ++ " " ++ show y | x:+y <- ps ]

File: acmOtherMaterials/Ulm Local Contest/2002/solution/diplomatic.cc
// Problem   Diplomatic License
// Algorithm Geometry
// Runtime   O(n)
// Author    Walter Guttmann
// Date      30.05.2002

#include <fstream>

#ifdef VERIFY
#define in cin
#else
ifstream in ("diplomatic.in");
#endif

int main ()
{
  int n;
  while (in >> n)
  {
    double x[n], y[n];
    for (int i=0 ; i<n ; i++)
      in >> x[i] >> y[i];
    cout << n;
    for (int i=0 ; i<n ; i++)
      cout.form (" %f %f", (x[i]+x[(i+1)%n])/2.0, (y[i]+y[(i+1)%n])/2.0);
    cout << endl;
  }
  return 0;
}

